sebs/knative/knative.py

```diff
@@ -0,0 +1,552 @@
+        ]
+
+        self.logging.info(f"Running build command: {' '.join(build_command)}")
+
+        try:
+            result = subprocess.run(
+                build_command,
+                capture_output=True,
+                check=True,
+            )
+            self.logging.info(f"Build output: {result.stdout.decode()}")
+        except subprocess.CalledProcessError as e:
+            self.logging.error(f"Error building the function: {e.stderr.decode()}")
+            raise RuntimeError(e) from e
+
+        self.logging.info(
+            f"Successfully built and pushed Docker image {repository_name}:{image_tag} "
+            f"to registry: {registry_name}."
+        )
+        return True
+
+    def package_code(
+        self,
+        directory: str,
+        language_name: str,
+        language_version: str,
+        benchmark: str,
+        is_cached: bool,
+    ) -> Tuple[str, int]:
+        """
+        Package code for Knative platform by building a Docker image.
+
+        Args:
+        - directory: Directory where the function code resides.
+        - language_name: Name of the programming language (e.g., Python).
+        - language_version: Version of the programming language.
+        - benchmark: Identifier for the benchmark or function.
+        - is_cached: Flag indicating if the code is cached.
+
+        Returns:
+        - Tuple containing the Docker image name (tag) and its size.
+        """
+
+        CONFIG_FILES = {
+            "python": ["func.py", "func.yaml", "Procfile", "requirements.txt"],
+            "nodejs": ["index.js", "func.yaml", "package.json"],
+        }
+
+        # Sanitize the benchmark name for the image tag
+        sanitized_benchmark_name = sanitize_benchmark_name_for_knative(benchmark)
+
+        # Generate func.yaml
+        func_yaml_content = {
+            "specVersion": "0.36.0",
+            "name": sanitized_benchmark_name,
+            "runtime": "node" if language_name == "nodejs" else language_name,
+            "created": datetime.now().astimezone().isoformat(),
+            "run": {
+                "envs": [
+                    {
+                        "name": "MINIO_STORAGE_CONNECTION_URL",
+                        "value": "{{ env:MINIO_STORAGE_CONNECTION_URL }}",
+                    },
+                    {
+                        "name": "MINIO_STORAGE_ACCESS_KEY",
+                        "value": "{{ env:MINIO_STORAGE_ACCESS_KEY }}",
+                    },
+                    {
+                        "name": "MINIO_STORAGE_SECRET_KEY",
+                        "value": "{{ env:MINIO_STORAGE_SECRET_KEY }}",
+                    },
+                ]
+            },
+        }
+        yaml_out = os.path.join(directory, "func.yaml")
+        with open(yaml_out, "w") as yaml_file:
+            yaml.dump(func_yaml_content, yaml_file, default_flow_style=False)
+
+        # Create Procfile for Python runtime
+        if language_name == "python":
```

Comment:
I think it would be nicer to extract them as separate functions and call them here - separation of concerns. One function for python mods, another for nodejs. It will be easier to extend it in the future :-)